dex/runner.go

```diff
@@ -86,41 +87,94 @@ type Connector interface {
 
 // ConnectionMaster manages a Connector.
 type ConnectionMaster struct {
-	contextManager
-	wg        *sync.WaitGroup
 	connector Connector
+	cancel    context.CancelFunc
+
+	// To permit calls to Done/On before Connect, guard the done channel.
+	mtx  sync.RWMutex
+	done chan struct{}
 }
 
-// NewConnectionMaster is the constructor for a new ConnectionMaster.
+// NewConnectionMaster creates a new ConnectionMaster. The Connect method should
+// be used before Disconnect. The On, Done, and Wait methods may be used at any
+// time. However, prior to Connect, Wait and Done immediately return and signal
+// completion, respectively.
 func NewConnectionMaster(c Connector) *ConnectionMaster {
 	return &ConnectionMaster{
 		connector: c,
 	}
 }
 
 // Connect connects the Connector, and returns any initial connection error. Use
-// Disconnect to shut down the Connector.
+// Disconnect to shut down the Connector. Connect is not safe for concurrent use
+// with Disconnect, and must be used before Disconnect.
 func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
-	c.init(ctx)
-	c.mtx.Lock()
-	c.wg, err = c.connector.Connect(c.ctx)
-	c.mtx.Unlock()
+	if c.On() { // probably a bug in the consumer
+		return errors.New("already running")
+	}
+
+	// Attempt to start the Connector.
+	ctx, cancel := context.WithCancel(ctx)
+	wg, err := c.connector.Connect(ctx)
 	if err != nil {
-		c.cancel() // otherwise On() says true
+		cancel() // no context leak
+		return err
 	}
-	return err
+
+	// It's running, enable Disconnect.
+	c.cancel = cancel // caller should synchronize Connect/Disconnect calls
+
+	// Done and On may be checked at any time.
+	done := make(chan struct{})
+	c.mtx.Lock()
+	c.done = done // after setting cancel
+	c.mtx.Unlock()
+
+	go func() { // capture the local variables
+		wg.Wait()
+		cancel() // if the Connector just died on its own, don't leak the context
+		close(done)
+	}()
```

Comment:
No way to avoid the goroutine? Probably doesn't matter though.

Ah. The goroutine gets rid of the `wg sync.WaitGroup` field.